rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,517 @@
+
+Where the service is a part of the visible counter name, and hostname and
+bucket must be stored and exported as a meta data.
+
+## App Telemetry Activation
+
+The collection and reporting of application telemetry should be activated by
+either of the following mechanisms:
+
+* presence of the `appTelemetry` property of the node in `nodexExt` entry of
+  the configuration (NOTE: the exact name and semantics of the value might be
+  defined by the server team later). In this case the SDK would either use
+  hardcoded path or extract one from the JSON configuration and use for and
+  build the endpoint address the websocket connection.
+
+* presence of the `app_telemetry_endpoint` property in the connection option,
+  or similar optiion in the `ClusterOptions`:
+
+      couchbases://abc.cloud.couchbase.com?app_telemetry_endpoint=ws://localhost:9090/app_telemetry
+
+      ClusterOptions options;
+      options.appTelemetryEndpoint("ws://localhost:9090/app_telemetry");
+
+  This endpoint must not have authentication, and should support WebSocket
+  protocol.
+
+## Network Interacation
+
+Once the SDK determined the endpoint it should establish HTTP connection and
+upgrade it to WebSocket.
+
+The server allows to close connection at any point of time.
+
+Once the server closes connection, the SDK should pick next available endpoint.
+Both initial and next endpoints must be randomized in case of more than one
+endpoint configured.
+
+The SDK must wait for backoff interval before connection attempt to avoid
+stressing the server. The SDK might use `app_telemetry_backoff` option to allow
+to change the backoff value.
+
+The SDK must either implement or enable WebSocket PING frame, and respond to
+server request with WebSocket PONG whenever the server requests it.
+
+## Wire Protocol
+
+The structure of the protocol is intentionally simple, and yet allows for
+future extension.
+
+The SDK leverages WebSocket framing, so does not need to have special fields
+for the payload size.
+
+First byte is reserved for Command codes in requests from the server, and for
+Status for SDK responses.
+
+This RFC defines the following types of the commands.
+
+| Name            | Value  | Description                                                                  |
+|-----------------|--------|------------------------------------------------------------------------------|
+| `GET_TELEMETRY` | `0x00` | Retrieve Application Telemetry from SDK in Prometheus Text Exposition Format |
+
+And the following statuses:
+
+| Name              | Value  | Description                                                |
+|-------------------|--------|------------------------------------------------------------|
+| `SUCCESS`         | `0x00` | No errors has been occured                                 |
+| `UNKNOWN_COMMAND` | `0x01` | The SDK does not support operation requested by the server |
+
+
+### `GET_TELEMETRY`
+
+The request must not have payload.
```

Comment:
Can we make this that the SDK should ignore any payload?  It lets us expand the protocol later and have existing SDK simply ignore things it doesn't understand - they just won't get sent in the response packets.

Reply:
Yes exactly. SDK_X could implement that active check, and that will prevent expanding GET_TELEMETRY down the line.

Reply:
I've changed `must not` to `should not`, so it does not look like the SDK have to implement the check